pip_run/scripts.py

```diff
@@ -73,7 +78,35 @@ def search(cls, params):
         return cls.try_read(next(files, None)).params()
 
     def read(self):
-        return self.read_comments() or self.read_python()
+        return self.read_toml() or self.read_comments() or self.read_python()
+
+    def read_toml(self):
+        r"""
+        >>> DepsReader('# /// script\n# dependencies = ["foo", "bar"]\n# ///\n').read()
+        ['foo', 'bar']
+        >>> DepsReader('# /// pyproject\n# dependencies = ["foo", "bar"]\n# ///\n').read_toml()
+        []
+        >>> DepsReader('# /// pyproject\n#dependencies = ["foo", "bar"]\n# ///\n').read_toml()
+        []
+        >>> DepsReader('# /// script\n# dependencies = ["foo", "bar"]\n').read_toml()
+        []
+        """
+        TOML_BLOCK_REGEX = r'(?m)^# /// (?P<type>[a-zA-Z0-9-]+)$\s(?P<content>(^#(| .*)$\s)+)^# ///$'
+        name = 'script'
+        matches = list(
+            filter(lambda m: m.group('type') == name, re.finditer(TOML_BLOCK_REGEX, self.script))
+        )
+        if len(matches) > 1:
+            raise ValueError(f'Multiple {name} blocks found')
+        elif len(matches) == 1:
+            content = ''.join(
+                line[2:] if line.startswith('# ') else line[1:]
+                for line in matches[0].group('content').splitlines(keepends=True)
+            )
+            deps = tomllib.loads(content).get("dependencies", [])
+        else:
+            deps = []
```

Comment:
The performance gain from `filter()` is already lost due to lambda. I'd suggest using a clean generator expression to enhance readability. And because only one match will be needed, we take only one. Next up, we check if there's another one and raise immediately, not knowing about any other possible matches as we don't need them.
It's also safer to use `.get("dependencies") or []` instead of `.get("dependencies", [])` in case `"dependencies"` is defined as null.
`line[2:] if line.startswith('# ') else line[1:]` can either be replaced with `line[2 if line.startswith('# ') else 1:]` or `line[1+line.startswith('# '):]`.

```suggestion
        iter_matches = (m for m in re.finditer(TOML_BLOCK_REGEX, self.script) if m.group('type') == name)
        match = next(iter_matches, None)
        if next(matches, None) is not None:
            raise ValueError(f'Multiple {name} blocks found')
        if match:
            content = ''.join(
                line[1+line.startswith('# '):]
                for line in match.group('content').splitlines(keepends=True)
            )
            deps = tomllib.loads(content).get("dependencies") or []
        else:
            deps = []
```

Reply:
This is taken straight from the PEP, where it's the canonical implementation of parsing. I'd rather *not* mess with it, simply because I don't see the benefit in risking the possibility that we introduce bugs by doing so.

Reply:
As the author, I am a strong -1 on this change because it is not as easy to translate into other languages as before.

Reply:
> As the author, I am a strong -1 on this change because it is not as easy to translate into other languages as before.

Thank you for your feedback. I was suggesting the change for pip-run specifically. Is there any other reason you don't like the suggestions?

Reply:
Nope sorry about that, as I mentioned in the edit to my comment feel free to do as you wish!

Reply:
I tested my changes against tests in this PR and observed no regression. In my opinion, a more detailed analysis of the change can disprove the concern regarding the risk of the bug.

Reply:
What is more, the suggested change costlessly handles behavior in an edge case not handled by the original version.

Reply:
Hi, sorry for joining an already crowded conversation, but @bswck asked for my thoughts, so here they are:

1. I do think it's worth checking that `dependencies` is a list of strings. Currently it looks like the user will get a confusing traceback if it's null or a number. Most importantly, if it's a single string, then each character will be treated as a dependency.
2. The rest of the suggestion doesn't seem to have any significant benefit and isn't worth this much discussion. If @pfmoore doesn't want to change this or talk about it then I support that, I'm sure he has plenty of things to do.
3. Checking the length of a list is vastly more readable than using `next/any`. I don't think it makes sense to worry about creating a big list of matches. If it does, I'd suggest using `islice` instead like [here](https://github.com/alexmojaki/executing/blob/6a54f9890d7f438b82647b42da2f3834f864f34f/executing/executing.py#L114-L119).
4. `deps[:] =` instead of `deps =` is jarring and weird to look at. And even if `deps[:] = ()` is faster than `deps = []` (is it?), even if the difference was significant (it obviously isn't), setting `deps = ()` would be fine anyway.
5. `line[1 + line.startswith('# '):]` is fun and clever but not super readable. `line[2 if line.startswith('# ') else 1:]` is maybe an improvement but it's subjective and insignificant.
6. I do prefer a list comprehension over filter/lambda, but :shrug:

Reply:
I'll see if I can find the time to review the various suggestions. TBH, this was very much a "drive by" PR, motivated by the fact that I use pip-run and I wanted to ensure that it supported the new standard. Copying the reference implementation was the fastest way to achieve that. I didn't want to spend a lot of time over details that could be hashed out later.

In many ways I'd argue that a robust PEP 723 parser should go in a library somewhere, although I can sympathise if people don't want a new dependency just for that one thing. I'm a little bit sad if we end up with lots of different implementations of the parsing code, all with their own quirks and trade-offs.

My recommendation would be that unless I get the time to do an update, the code can go in as is, and it can be fixed in a follow-up PR. I'm absolutely *not* going to get upset about someone updating the code in a follow-up.